programs/lib/startup.c

```diff
@@ -52,35 +66,67 @@ void free(void *ptr) {
  * Initialize initialized global variables, set uninitialized global variables
  * to zero, configure tinyalloc, and jump to main.
  */
-void Reset_Handler(void) {
-    // Copy .data section into the RAM
-    uint32_t size   = &__data_end__ - &__data_start__;
-    uint32_t *pDst  = (uint32_t*)&__data_start__;       // RAM
-    uint32_t *pSrc  = (uint32_t*)&__etext;              // ROM
-
-    for (uint32_t i = 0; i < size; i++) {
-        *pDst++ = *pSrc++;
-    }
+void Reset_Handler() {
+    // Get hartid
+    uint32_t hartid = read_hartid();
 
-    // Init. the .bss section to zero in RAM
-    size = (uint32_t)&__bss_end__ - (uint32_t)&__bss_start__;
-    pDst = (uint32_t*)&__bss_start__;
-    for(uint32_t i = 0; i < size; i++) {
-        *pDst++ = 0;
-    }
+    // Only thread 0 performs the setup,
+    // the other threads busy wait until ready.
+    if (hartid == 0) {
+        // Copy .data section into the RAM
+        uint32_t size   = &__data_end__ - &__data_start__;
+        uint32_t *pDst  = (uint32_t*)&__data_start__;       // RAM
+        uint32_t *pSrc  = (uint32_t*)&__etext;              // ROM
+
+        for (uint32_t i = 0; i < size; i++) {
+            *pDst++ = *pSrc++;
+        }
 
-    // Initialize tinyalloc.
-    ta_init( 
-        &end, // start of the heap space
-        DSPM_LIMIT,
-        TA_MAX_HEAP_BLOCK, 
-        16, // split_thresh: 16 bytes (Only used when reusing blocks.)
-        TA_ALIGNMENT
-    );
+        // Init. the .bss section to zero in RAM
+        size = (uint32_t)&__bss_end__ - (uint32_t)&__bss_start__;
+        pDst = (uint32_t*)&__bss_start__;
+        for(uint32_t i = 0; i < size; i++) {
+            *pDst++ = 0;
+        }
+
+        // Initialize tinyalloc.
+        ta_init( 
+            &end, // start of the heap space
+            DSPM_LIMIT,
+            TA_MAX_HEAP_BLOCK, 
+            16, // split_thresh: 16 bytes (Only used when reusing blocks.)
+            TA_ALIGNMENT
+        );
+
+        // Signal ready.
+        hwlock_acquire();
+        __ready__ = true;
+        hwlock_release();
+    } else {
+        // Wait for thread 0 to finish setup.
+        // FIXME: Use delay until (DU)
+        // for precise synchronization.
+        while (!__ready__);
+    }
 
     // Call main().
-    main();
+    if (hartid == 0) {
+        main();
+    } else {
+        worker_main();
+    }
 
     // Exit by calling the _exit() syscall.
-    _exit(0);
+    if (hartid == 0) {
+        // Mark the main thread as exited.
+        num_threads_exited += 1;
```

Comment:
I didnt see where the other harts increment this global variable. Also, is there a race condition if multiple threads try to increment the same global variable without a lock?

Reply:
The other harts increment the variable in their `worker_main`. The race condition is a good catch. Will fix now.